File: sdk/jvm/src/main/java/com/dansherwin/devlogbus/DevLogBusLogger.java
package com.dansherwin.devlogbus;

import java.io.IOException;

public final class DevLogBusLogger {
    private final DevLogBusClient client;

    DevLogBusLogger(DevLogBusClient client) {
        this.client = client;
    }

    public DevLogBusClient.PublishResult debug(String message, String attrsJson) throws IOException, InterruptedException {
        return client.publishMessage("DEBUG", message, attrsJson);
    }

    public DevLogBusClient.PublishResult info(String message, String attrsJson) throws IOException, InterruptedException {
        return client.publishMessage("INFO", message, attrsJson);
    }

    public DevLogBusClient.PublishResult warn(String message, String attrsJson) throws IOException, InterruptedException {
        return client.publishMessage("WARN", message, attrsJson);
    }

    public DevLogBusClient.PublishResult error(String message, String attrsJson) throws IOException, InterruptedException {
        return client.publishMessage("ERROR", message, attrsJson);
    }
}

File: sdk/dotnet/DevLogBus.Sdk/DevLogBusLogger.cs
namespace DanSherwin.DevLogBus;

public sealed class DevLogBusLogger
{
    private readonly DevLogBusClient client;
    private readonly string source;

    internal DevLogBusLogger(DevLogBusClient client, string source)
    {
        this.client = client;
        this.source = source;
    }

    public Task<DevLogBusPublishResult> DebugAsync(
        string message,
        IReadOnlyDictionary<string, object?>? attrs = null,
        CancellationToken cancellationToken = default)
    {
        return PublishAsync("DEBUG", message, attrs, cancellationToken);
    }

    public Task<DevLogBusPublishResult> InfoAsync(
        string message,
        IReadOnlyDictionary<string, object?>? attrs = null,
        CancellationToken cancellationToken = default)
    {
        return PublishAsync("INFO", message, attrs, cancellationToken);
    }

    public Task<DevLogBusPublishResult> WarnAsync(
        string message,
        IReadOnlyDictionary<string, object?>? attrs = null,
        CancellationToken cancellationToken = default)
    {
        return PublishAsync("WARN", message, attrs, cancellationToken);
    }

    public Task<DevLogBusPublishResult> ErrorAsync(
        string message,
        IReadOnlyDictionary<string, object?>? attrs = null,
        CancellationToken cancellationToken = default)
    {
        return PublishAsync("ERROR", message, attrs, cancellationToken);
    }

    private Task<DevLogBusPublishResult> PublishAsync(
        string level,
        string message,
        IReadOnlyDictionary<string, object?>? attrs,
        CancellationToken cancellationToken)
    {
        return client.PublishAsync(new DevLogBusRecord
        {
            Source = source,
            Level = level,
            Message = message,
            Attrs = attrs,
        }, cancellationToken);
    }
}
